src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1571,6 +1571,180 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
+      jccb(Assembler::carryClear, *larr[i]);
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+      bind(*larr[i]);
+      incq(midx);
+    }
+  }
+}
+#endif // _LP64
+
+void C2_MacroAssembler::vgather8b(BasicType elem_bt, XMMRegister dst,
+                                  Register base, Register idx_base,
+                                  Register rtmp, int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      // dst[i] = src[index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      // dst[i] = src[index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+void C2_MacroAssembler::vgather8b_offset(BasicType elem_bt, XMMRegister dst,
+                                         Register base, Register idx_base,
+                                         Register offset, Register rtmp,
+                                         int vlen_enc) {
+  vpxor(dst, dst, dst, vlen_enc);
+  if (elem_bt == T_SHORT) {
+    for (int i = 0; i < 4; i++) {
+      // dst[i] = src[offset + index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrw(dst, Address(base, rtmp, Address::times_2), i);
+    }
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    for (int i = 0; i < 8; i++) {
+      // dst[i] = src[offset + index[i]]
+      movl(rtmp, Address(idx_base, i * 4));
+      addl(rtmp, offset);
+      pinsrb(dst, Address(base, rtmp), i);
+    }
+  }
+}
+
+/*
+ * Gather loop first packs 4 short / 8 byte values from gather indices
+ * into quadword lane and then permutes quadword lane into appropriate
+ * location in destination vector. Following pseudo code describes the
+ * algorithm in detail:-
+ *
+ * DST_VEC = ZERO_VEC
+ * PERM_INDEX = {0, 1, 2, 3, 4, 5, 6, 7, 8..}
+ * TWO_VEC    = {2, 2, 2, 2, 2, 2, 2, 2, 2..}
+ * FOREACH_ITER:
+ *     TMP_VEC_64 = PICK_SUB_WORDS_FROM_GATHER_INDICES
+ *     TEMP_PERM_VEC = PERMUTE TMP_VEC_64 PERM_INDEX
+ *     DST_VEC = DST_VEC OR TEMP_PERM_VEC
+ *     PERM_INDEX = PERM_INDEX - TWO_VEC
+ *
+ * With each iteration, doubleword permute indices (0,1) corresponding
+ * to assembled quadword gets right shifted by two lane position.
+ *
+ */
+void C2_MacroAssembler::vgather_subword(BasicType elem_ty, XMMRegister dst,
+                                        Register base, Register idx_base,
+                                        Register offset, Register mask,
+                                        XMMRegister xtmp1, XMMRegister xtmp2,
+                                        XMMRegister xtmp3, Register rtmp,
+                                        Register midx, Register length,
+                                        int vector_len, int vlen_enc) {
```

Comment:
I would like to see more descriptive names, where I don't have to reverse-engineer their meaning.
What are the pre/post-conditions on `midx`?

Reply:
I'll rereview after

Reply:
So `xtmp1...3` and `rtmp` cannot have more descriptive names?